// welly/lib/welly.js
/*
	author: wangningbo
	data:2013/11/17
	modify-log:
	-----------------------------------------------------------------------------------------
	|	NAME 		DATE		LOG 														|
	-----------------------------------------------------------------------------------------
*/


var utils = require('../utils');
var welly = {
	version: '1.0.1',
	author: 'welly-company',
	router: require('router'),
	middleware: require('middleware'),
	controller: require('controller'),
	config:{
		'views' : '/views',
		'controllers' : '/controllers'
	}
};

var set = exports.set = function(key,value){
	if (key === 'api') {
		welly.controller.setApiPath(value);
	};
	if (key === 'views') {
		welly.controller.setViewPath(value);
	};
	if (key === 'view engine') {
		welly.controller.setViewEngine(value);
	};
	welly.config[key] = value || welly.config[key];
};
set('view engine',require('ejs'));

exports.use = function(module_name, module){
	welly[util.toCamule(module_name)] = module || require(module_name);
};

exports.add = welly.middleware.add;

exports.process = function(req,res){
	check_cfg();
	var req = welly.middleware.wrapRequest(req),
		res = welly.middleware.wrapResponse(res);
	welly.router.route(welly.config.controllers,welly.config.controllers,req,res);
};

exports.controller = welly.controller;
exports.utils = utils;

function check_cfg(){
	if (!welly.config['view engine']) {
		throw '================================================================='+
			'\nMiss view engine.'+
			'\nplease add:'+
				'\n\twelly.use("views engine",your_view_engine_module);'+
			'\nto your start-server-script.'+
			'\nFor example:'+
			'\n\twelly.use("views engine",require("ejs"));'+
			'\n=================================================================';
	};
	if (!welly.config['views']) {
		throw '================================================================='+
			'\nMiss views path config.'+
			'\nplease add:'+
				'\n\twelly.set("views","your_views_folder_path");'+
			'\nto your start-server-script.'+
			'\n=================================================================';
	};
}
