// middleware/index.js
/*
	wrap request and response for better using
	author: wangningbo
	data:2013/11/17
	modify-log:
	-----------------------------------------------------------------------------------------
		NAME 		DATE		LOG
		wangningbo	2013/11/17	add cookie wrapper
		wangningbo	2013/11/17	add session wrapper
	-----------------------------------------------------------------------------------------
*/

var middleware = {
	req: [],
	res: []
};

exports.wrapRequest = function(req){
	for (var i = 0; i < middleware.req.length; i++) {
		middleware.req[i].process(req);
	};
	return req;
};
exports.wrapResponse = function(res){
	for (var i = 0; i < middleware.res.length; i++) {
		middleware.res[i].process(res);
	};
	return res;
};
exports.add = function(name,middleware,type){
	switch(type){
		case undefined:
			middleware.req.push(middleware);
			middleware.res.push(middleware);
			break;
		case 'request':
			middleware.req.push(middleware);
			break;
		case 'response':
			middleware.res.push(middleware);
			break;
		default:
			break;
	}
};
