// controller/lib/proto.js
var http = require('http');

module.exports = {
	__setup: function(req,res){
		this.req = req;
		this.res = res;
	},
	render: function(path,data){
		var self = this;
		this.config.viewEngine.renderFile(this.config.view+path+'.html',data,function(err,str){
			self.end(str);
		});
	},
	end: function(str){
		this.res.end(str);
	},
	_onFinish: function(key,data){
		this.data[key] = JSON.parse(data);
		this.count--;
		
		if (this.count === 0) {
			this.emit('end',this.data);
		};
	},
	_send: function(url,key){
		var self = this,
			url = url.split('::'),
			host = this.config.api[url[0]];
			path = url[1];
		if (!path) {
			console.log('host ' + host + ' is missing');
			this.count--;
			return ;
		};
		var req = http.request(
			{
				hostname:host,
				path:path,
				method: this.req.method,
				port:80
			},
			function(res){
				res.on('data',function(data){
					self.emit('data',key,data.toString());
				});
			}
		);
		req.on('error',function(error){
			console.log('Request Error: | path:'+url+' | '+error);
			self.emit('error',key);
		});
		req.end();
	},
	send: function(interfaces){
		this.count = 0;
		this.on('data',this._onFinish);
		this.on('error',this._onFinish);
		for(var i in interfaces){
			this.count++;
			this._send(interfaces[i],i);
		}
	}
}
